index.rst

```diff
@@ -1,61 +1,461 @@
+
+
+EFD client
+==========
+
+- Show that missing values in the InfluxDB are converted back to NaN when the EFD client returns a Pandas dataframe.
+
+- Point to the documentation on how to handle missing data in Pandas, and say why NaNs are more convenient than convenient
+
+- Point out that Pandas can represent missing values for all data types
+
+- Make sure the EFD Client reconstruct arrays with missing values correctly
+
+
+Conclusion
+==========
 
-   **This technote is not yet published.**
+We reviewed how DDS, Kafka, and InfluxDB handle missing values.
+In the ideal world, every field in a topic should be optional and have a default value that represents a missing value for each data type.
+However, adding support to optional fields at this point represents a significant effort, especially to make the CSCs treat optional fields properly.
+It also implies changing SAL Kafka, the InfluxDB Sink connector, and potentially the ``aioinflux`` client to get everything right.
 
-   Capture CAP meeting discussions and final decision on how to represent missing values in the EFD.
+Using a sentinel value like ``NaN`` to represent missing values works for ``float`` and ``double`` IDL types as a partial solution.
+A field with the value ``NaN`` in a DDS record is converted to a field with value ``NaN`` in the Avro record,  and it is dropped by the InfluxDB Sink connector.
+In particular, we showed that when querying InfluxDB in Python, it fills the missing values with ``None``.
+When using the EFD client, missing values are converted back to ``NaN`` which is a convenient way to represent missing values with Pandas ``float64`` dtype.
 
-.. Add content here.
-.. Do not include the document title (it's automatically added from metadata.yaml).
+This solution is currently in place, and it solves our problem for about 80% of the fields in telemetry topics. Our recommendation for DDS topics with sparse arrays of ``float`` or ``double`` IDL types is to fill in the missing values also with ``NaN`` values.
 
-.. .. rubric:: References
+We don't have a solution for representing missing values for other data types yet. If there was, for example, an obvious symbol to represent missing values for integers, we could change the InfluxDB Sink connector to drop the fields with that symbol. InfluxDB would denote the value as missing, and when the InfluxDB client converts the result to Pandas we could use the `Pandas nullable integer data type`_.
 
-.. Make in-text citations with: :cite:`bibkey`.
+.. _Pandas nullable integer data type: https://pandas.pydata.org/pandas-docs/stable/user_guide/integer_na.html
 
-.. .. bibliography:: local.bib lsstbib/books.bib lsstbib/lsst.bib lsstbib/lsst-dm.bib lsstbib/refs.bib lsstbib/refs_ads.bib
-..    :style: lsst_aa
+As a final recommendation, we should consider using the ``@default`` annotation in the IDL schema to set the default values for ``float`` and ``double`` fields to ``NaN`` instead of ``0``.
```

Comment:
Oh, I think this should be much more front and center.  I didn't appreciate that 0 was the default for float and double until I got to here!  Maybe we should put a bullet list of recommendations right at the beginning.

Reply:
Agreed. I'll add a new section for recomendation to make this more explicit.